### PR TITLE
PFAUnit implement IERC20Payable

### PR DESCRIPTION
**Pull request by xiangformless**
- setup IERC20Payable interface for PFAUnit
- create unit test with truffle

### CODE REVIEW

**Comment by xiangformless** [comment]
image
successful local truffle test

**Comment by kim-formless** [comment]
> image successful local truffle test

@xiangformless nice!!

**Comment by un7c0rn** [comment]
> image successful local truffle test

Great progress @xiangformless!

**Review by un7c0rn — CHANGES_REQUESTED** [review]
@xiangformless PTAL at test failure in presubmits.

File: contracts/PFAUnit.sol [hunk 1]
```diff
@@ -11,27 +11,30 @@
...
     /// PFA.
     event PaymentToOwner(address indexed owner, uint256 value);
 
     string public constant NAME = "SHARE";
     string public constant SYMBOL = "PFA";
     uint256 private constant UNIT_TOKEN_INDEX = 0;
-
     string internal _tokenURI;
 
+    // ERC-20 contract address e.g. USDC
```
Comment (un7c0rn) [inline contracts/PFAUnit.sol:34]: nit: s/ERC-20/ERC20

File: contracts/PFAUnit.sol [hunk 2]
```diff
@@ -11,27 +11,30 @@
...
+    address private erc20ContractAddress;
```
Comment (un7c0rn) [inline contracts/PFAUnit.sol]: nit: By convention in our codebase, for private Solidity variables let's use the underscore prefix, eg. `_erc20ContractAddress`.

File: contracts/PFAUnit.sol (hunk 1 above)
Comment (un7c0rn) [reply contracts/PFAUnit.sol:34]: Here and elsewhere that applies.

File: contracts/PFAUnit.sol [hunk 3]
```diff
@@ -104,12 +107,24 @@ contract PFAUnit is PFA, ERC721 /* G_NFT */ {
...
+
+    /// @notice Gets the ERC-20 contract address used for payments.
+    function getERC20ContractAddress() external view override(IERC20Payable) returns (address) {
+        return erc20ContractAddress;
+    }
+
     function supportsInterface(
         bytes4 interfaceId
-    ) public view virtual override returns (bool) {
+    ) public view virtual override(ERC721) returns (bool) {
```
Comment (un7c0rn) [inline contracts/PFAUnit.sol:123]: Is `override(ERC721)` needed here?

File: contracts/PFAUnit.sol [hunk 4]
```diff
@@ -104,12 +107,24 @@ contract PFAUnit is PFA, ERC721 /* G_NFT */ {
...
     }
 
+    /// @notice Sets the ERC-20 contract address (e.g., for USDC payments).
+    function setERC20ContractAddress(address contractAddress_) external override onlyOwner {
+        require(contractAddress_ != address(0), "Invalid ERC20 contract address");
+        erc20ContractAddress = contractAddress_;
+    }
+
+    /// @notice Gets the ERC-20 contract address used for payments.
+    function getERC20ContractAddress() external view override(IERC20Payable) returns (address) {
```
Comment (un7c0rn) [inline contracts/PFAUnit.sol]: Is `override(IERC20Payable)` needed here?

File: contracts/interfaces/IERC20Payable.sol [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+// ⣿⣿⠀⠀⠀⠀⠀⢿⣿⣤⣾⣿⠇⠀⠀⣿⣿⠀⣿⣿⠀⠀⠀⣿⣿⠀⣿⡇⠈⣿⡇⠀⠀⣿⣿⣤⣤⡄⠀⢸⣿⣧⣤⣤⡄⠀⢸⣿⣆⠀⣿⣿⠀⠀⣿⣿⡀⢀⣿⣿
+// ⠛⠛⠀⠀⠀⠀⠀⠈⠛⠿⠿⠛⠀⠀⠀⠛⠛⠀⠘⠛⠃⠀⠀⠛⠛⠀⠛⠀⠈⠛⠃⠀⠀⠛⠛⠛⠛⠃⠀⠘⠛⠛⠛⠛⠃⠀⠀⠙⠿⠿⠟⠁⠀⠀⠀⠛⠿⠿⠛⠀
+// https://formless.xyz/opportunities
+//⠀
+pragma solidity >=0.8.0 <0.9.0;
+
+/// @title Contract which implements it can receive payment using ERC20 tokens including USDC
+/// @author xiang@formless.xyz
+interface IERC20Payable {
+    function setERC20ContractAddress(address contractAddress_) external;
```
Comment (un7c0rn) [inline contracts/interfaces/IERC20Payable.sol:16]: nit: Spacing

File: contracts/interfaces/IERC20Payable.sol [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+   function getERC20ContractAddress() external view returns (address);
+}
```
Comment (un7c0rn) [inline contracts/interfaces/IERC20Payable.sol:18]: nit: Newline

File: contracts/mocks/MockUSDC.sol [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+// ⣿⣿⠀⠀⠀⠀⠀⢿⣿⣤⣾⣿⠇⠀⠀⣿⣿⠀⣿⣿⠀⠀⠀⣿⣿⠀⣿⡇⠈⣿⡇⠀⠀⣿⣿⣤⣤⡄⠀⢸⣿⣧⣤⣤⡄⠀⢸⣿⣆⠀⣿⣿⠀⠀⣿⣿⡀⢀⣿⣿
+// ⠛⠛⠀⠀⠀⠀⠀⠈⠛⠿⠿⠛⠀⠀⠀⠛⠛⠀⠘⠛⠃⠀⠀⠛⠛⠀⠛⠀⠈⠛⠃⠀⠀⠛⠛⠛⠛⠃⠀⠘⠛⠛⠛⠛⠃⠀⠀⠙⠿⠿⠟⠁⠀⠀⠀⠛⠿⠿⠛⠀
+// https://formless.xyz/opportunities
+//
+pragma solidity >=0.8.0 <0.9.0;
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+/// @title Mock Liquidity Pool.
+/// For testing purposes only.
+/// @author brandon@formless.xyz
```
Comment (un7c0rn) [inline contracts/mocks/MockUSDC.sol]: s/brandon/xiang

File: package.json [hunk 8]
```diff
@@ -5,9 +5,10 @@
   "author": "brandon@formless.xyz",
   "license": "UNLICENSED",
   "dependencies": {
-    "@openzeppelin/contracts": "^4.4.2",
+    "@openzeppelin/contracts": "^4.9.6",
```
Comment (un7c0rn) [inline package.json]: nit: Let's leave this at the prior version and pin. The reason is these impact the contract build sizes when we init the submodule in our internal BE, and will cause test failures until we synchronize the new dependencies.

File: package.json [hunk 9]
```diff
@@ -5,9 +5,10 @@
...
     "@truffle/hdwallet-provider": "^2.0.0",
     "ethereumjs-util": "^7.1.4",
+    "mocha": "^10.8.2",
```
Comment (un7c0rn) [inline package.json:11]: Similarly here let's actually pin all versions.

**Review by un7c0rn — CHANGES_REQUESTED** [review]
Thanks @xiangformless, looks great! Notes inline.

File: contracts/PFAUnit.sol (hunk 3 above)
Comment (xiangformless) [reply contracts/PFAUnit.sol:123]: yes, without it, there is an error in this supportsInterface function.

File: contracts/PFAUnit.sol (hunk 4 above)
Comment (xiangformless) [reply contracts/PFAUnit.sol]: removed

File: package.json [hunk 10]
```diff
@@ -8,6 +8,7 @@
     "@openzeppelin/contracts": "^4.4.2",
```
Comment (un7c0rn) [inline package.json:8]: @xiangformless Can you pin these other packages as well? eg. remove the ^

Comment (xiangformless) [reply package.json:8]: image
is this OK? or pin all of them?

Comment (xiangformless) [reply package.json:8]: @un7c0rn

**Comment by xiangformless** [comment]
@un7c0rn this test keeps failing. It does not seem to be caused by my code. Could you please take a look? thank you
image

**Comment by un7c0rn** [comment]
@xiangformless Yup taking a look now.

**Comment by un7c0rn** [comment]
@xiangformless Passes locally for me as well.

**Comment by un7c0rn** [comment]
@xiangformless It could be the package changes I requested. Can you try restoring package.json to match what's in `main` apart from the changes you need? Apologies.

**Review by un7c0rn — APPROVED** [review]
LGTM

**Pull request merged**